include/crypto/evp.h

```diff
@@ -746,6 +750,15 @@ struct evp_pkey_st {
 # define EVP_PKEY_CTX_IS_KEM_OP(ctx) \
     (((ctx)->operation & EVP_PKEY_OP_TYPE_KEM) != 0)
 
+struct evp_skey_st {
+    /* == Common attributes == */
+    CRYPTO_REF_COUNT references;
+    CRYPTO_RWLOCK *lock;
+
+    void *keydata; /* Alg-specific key data */
+    EVP_KEYMGMT *keymgmt; /* Mostly for freeing keydata */
+}; /* EVP_SKEY */
```

Comment:
Add a "bunch of bytes" pointer and length.  Something to realise is that *if* that pointer is non-NULL, the `keymgmt` pointer may be NULL.

(actually, the `keydata` pointer could be used for the "bunch of bytes" as well, and be understood as such as long as `keymgmt == NULL`)

Reply:
With that, libcrypto functions that support EVP_SKEY can fall back to use provider functions that use bunch of bytes form of key (i.e. the non-skey functions).

Reply:
Once we have a "default provider" keydata structure you'll be able to reuse it to support providers that do not have skey mgmt, by exporting from default provider's format via EVP_SKEY_todata() before initializing  a legacy init() cipher function.

Key export/import, just like it is done for Public Keys will be a reality in some scenarios, so it make sense to just reuse as much as possible the same code paths and reduce "special" stuff.

Reply:
I am not quite sure this is a good idea. What we envisioned at the OTC meeting were potentially multiple keydata representations of the SKEY inside multiple providers. For example the providers could be able to transfer a wrapped key if they somehow shared the wrapping key before. However this might be implemented later.

In a first iteration I'd like to see a _bunch of bytes_ import and export functions - of course they might be optional and not implemented either if the given provider does not support exporting raw secret keys. So the "bunch of bytes" representation will be optional as well.

The lifecycle of an SKEY would be:
1. generation by a KDF or KEX or KEM or initialization from an application-provided bunch of bytes or by getting a reference as a result of a lookup by OSSL_STORE.
2. usage of the key in various new APIs which would allow passing an SKEY - this might trigger an import or export + import of the bunch of bytes data, which might of course fail for keys that are not exportable or if the provider selected for the operation does not allow importing keys - it would be useful to keep the references in a cache to avoid doing the import again for another operation with the SKEY in the same provider
3. release of the SKEY - with key being deleted from the provider if that key was not a permanent key (or was just imported there for an operation)

I realize it might be needed to have a type associated with the SKEY but that could be just an additional OSSL_PARAM.

Reply:
@t8m for the record I think we are just saying the same thing, my idea is that the default provider keydata type can simply be a struct that just has a pointer to OSSL_PARAM for quick import/export to/from other providers, after all the default provider will also have to support APIs that use an SKEY so it will need its own representation as keydata as well